### PR TITLE
A few follow-ups for taproot signing

### PR DESCRIPTION
**Pull request by sipa**
This addresses a few review comments from #21365 that were left at the time of merge (as well as some from #22166 applying to the commit it shared with #21365).

I do not think any are blockers for a 22.0 release.

### CODE REVIEW

**Comment by theStack** [comment]
Concept ACK

**Comment by achow101** [comment]
aCK 38844bc18492a847a6a3aff000eb0e85571070dd

**Comment by sipa** [comment]
Improved the comment in `CKey::SignSchnorr` to elaborate on the 3 variants of tweaking (pointed out [here](https://github.com/bitcoin/bitcoin/pull/21365#discussion_r653907350)).

**Comment by Sjors** [comment]
utACK 9336670

**Review by theStack — APPROVED** [review]
ACK 9336670df5e0b8baa4169f13ba1d242ecae85d26

Checked that the introduced comments match the logic, verified by review that the code-changes are refactor-only (+ an added sanity check in `TaprootBuilder::GetSpendData`), built locally, ran unit tests and the functional test `feature_taproot.py`.

File: src/script/sign.cpp [hunk 1]
```diff
@@ -636,21 +636,18 @@ bool SignTransaction(CMutableTransaction& mtx, const SigningProvider* keystore,
...
     std::vector<CTxOut> spent_outputs;
-    spent_outputs.resize(mtx.vin.size());
-    bool have_all_spent_outputs = true;
     for (unsigned int i = 0; i < mtx.vin.size(); i++) {
         CTxIn& txin = mtx.vin[i];
         auto coin = coins.find(txin.prevout);
         if (coin == coins.end() || coin->second.IsSpent()) {
-            have_all_spent_outputs = false;
+            txdata.Init(txConst, {}, true);
+            break;
```
Comment (jonatack) [inline src/script/sign.cpp:644]: 2b19345 suggestion
```diff
@@ -636,11 +636,11 @@ bool SignTransaction(CMutableTransaction& mtx, const SigningProvider* keystore,
 
     PrecomputedTransactionData txdata;
     std::vector<CTxOut> spent_outputs;
-    for (unsigned int i = 0; i < mtx.vin.size(); i++) {
+    for (unsigned int i = 0; i < mtx.vin.size(); ++i) {
         CTxIn& txin = mtx.vin[i];
         auto coin = coins.find(txin.prevout);
         if (coin == coins.end() || coin->second.IsSpent()) {
-            txdata.Init(txConst, {}, true);
+            txdata.Init(txConst, /* spent_outputs */ {}, /* force */ true);
             break;
         } else {
             spent_outputs.emplace_back(coin->second.out.nValue, coin->second.out.scriptPubKey);
@@ -651,7 +651,7 @@ bool SignTransaction(CMutableTransaction& mtx, const SigningProvider* keystore,
     }
 
     // Sign what we can:
-    for (unsigned int i = 0; i < mtx.vin.size(); i++) {
+    for (unsigned int i = 0; i < mtx.vin.size(); ++i) {
         CTxIn& txin = mtx.vin[i];
```

File: src/script/interpreter.h [hunk 2]
```diff
@@ -167,6 +167,13 @@ struct PrecomputedTransactionData
 
     PrecomputedTransactionData() = default;
 
+    /** Initialize this PrecomputedTransactionData with transaction data.
+     *
+     * @param[in]   tx             The transaction for which data is being precomputed.
+     * @param[in]   spent_outputs  The CTxOuts being spent, one for each tx.vin, in order
+     * @param[in]   force          Precompute data for all optional features, regardless of
+     *                             what is in the inputs (used at signing time, when the inputs
+     *                             aren't filled in yet). */
```
Comment (jonatack) [inline src/script/interpreter.h:176]: 8b1d3db suggestion
```diff
@@ -170,10 +170,10 @@ struct PrecomputedTransactionData
     /** Initialize this PrecomputedTransactionData with transaction data.
      *
      * @param[in]   tx             The transaction for which data is being precomputed.
-     * @param[in]   spent_outputs  The CTxOuts being spent, one for each tx.vin, in order
-     * @param[in]   force          Precompute data for all optional features, regardless of
-     *                             what is in the inputs (used at signing time, when the inputs
-     *                             aren't filled in yet). */
+     * @param[in]   spent_outputs  The CTxOuts being spent, one for each tx.vin, in order.
+     * @param[in]   force          Whether to precompute data for all optional features,
+     *                             regardless of what is in the inputs (used at signing
+     *                             time, when the inputs aren't filled in yet). */
```

**Review by jonatack — COMMENTED** [review]
ACK 9336670df5e0b8baa4169f13ba1d242ecae85d26

File: src/script/interpreter.h (hunk 2 above)
Comment (sipa) [reply src/script/interpreter.h:176]: Done.

File: src/script/sign.cpp (hunk 1 above)
Comment (sipa) [reply src/script/sign.cpp:644]: Done.

**Review by theStack — COMMENTED** [review]
It seems like after rebasing, the `CKey::SignSchnorr` comment improvement (previously done in commit https://github.com/bitcoin/bitcoin/commit/8b1d3dbfac899c19ab92e3591157d9f4a8c3ca49) is missing now?

**Review by jonatack — COMMENTED** [review]
ACK per `git range-diff 7a49fdc 9336670 f0fb5f8` modulo the missing `CKey::SignSchnorr()` doxygen improvement as mentioned in https://github.com/bitcoin/bitcoin/pull/22275#pullrequestreview-698581320

**Comment by sipa** [comment]
@theStack @jonatack Restored the accidentally deleted comment improvement.

**Review by theStack — APPROVED** [review]
re-ACK 08f57a0057c58f19cd8ae3de89548d014980478a 🌴 

Checked via `git range-diff 9336670d...08f57a00` that since my last ACK, the only changes were jonatack's suggestions w.r.t. to doxygen comments wording and code style (https://github.com/bitcoin/bitcoin/pull/22275#discussion_r659299734, https://github.com/bitcoin/bitcoin/pull/22275#discussion_r659300047) plus the rebase on master.

**Review by Zero-1729 — COMMENTED** [review]
crACK 08f57a0 🧉

LGTM, changes check out and taproot tests passed locally.

**Review by jonatack — COMMENTED** [review]
Code review ACK 08f57a0057c58f19cd8ae3de89548d014980478a per `git range-diff e9d6eb1 9336670 08f57a0` followed by re-code review per commit to swap context back into memory and debug build/run unit tests + feature_taproot.py as a sanity check

**Review by uvhw — COMMENTED** [review]
#14529

**Pull request merged**